Add workload resolver

dsplaisted opened the pull request:
Fix #11228
Fix #11229

dsplaisted committed:
Add workload SDK resolver project

dsplaisted committed:
Import workload props and targets

dsplaisted committed:
Resolve workload manifests and packs from dotnet root

dsplaisted committed:
Resolve workload manifests on .NET Framework

dsplaisted committed:
Read workload manifest from XML

dsplaisted committed:
Include workload resolver in MSBuildSdkResolver

dsplaisted committed:
Update to new workload manifest reader APIs, compile with MSBuild SDK resolver

dsplaisted committed:
Switch to sdk-manifests folder

dsplaisted committed:
Don't crash on workload manifest folder without manifest

dsplaisted committed:
Add test workloads to layout

dsplaisted committed:
Add workload tests

wli3 commented on src/Tasks/Microsoft.NET.Build.Tasks/targets/Microsoft.NET.Sdk.targets:
```diff
@@ -1010,6 +1010,36 @@ Copyright (c) .NET Foundation. All rights reserved.
+    <!--<WorkloadManifestRoot Condition="'$(WorkloadManifestRoot)' == ''">C:\git\msbuild-sdk-resolver-test\testresolver\sdk-manifests\$(SdkFeatureBand)</WorkloadManifestRoot>-->
+   
+    <!-- TODO: This should probably go in BundledVersions.props-->
+    <WorkloadManifestRoot Condition="'$(WorkloadManifestRoot)' == ''">$([MSBuild]::EnsureTrailingSlash('$(NetCoreRoot)'))sdk-manifests\$(SdkFeatureBand)</WorkloadManifestRoot>
```
SdkFeatureBand and the ManifestRoot is calcuated in manifest reader too. We should only keep one logic

dsplaisted replied:
This is addressed now by using a new pseudo-SDK (Microsoft.NET.SDK.WorkloadManifestTargetsLocator) so the resolver can return the right paths.

wli3 commented on src/Layout/redist/redist.csproj:
```diff
@@ -48,6 +48,7 @@
                       ReferenceOutputAssembly="false" />
     <ProjectReference Include="..\..\Tasks\Microsoft.NET.Build.Extensions.Tasks\Microsoft.NET.Build.Extensions.Tasks.csproj"
                       ReferenceOutputAssembly="false" />
+    <ProjectReference Include="..\..\Resolvers\Microsoft.NET.Sdk.WorkloadResolver\Microsoft.NET.Sdk.WorkloadResolver.csproj" />
```
why this is needed? It is not part of existing msbuild resolver's dependency?

dsplaisted replied:
The workload resolver is combined with the SDK resolver for the full MSBuild resolver.  However, that resolver isn't used at all in .NET Core MSBuild, as there it doesn't need to go through the global.json resolution.  So the workload resolver is still in a separate, new MSBuild resolver DLL for .NET Core.

wli3 replied:
Could you test both on commandline and in VS? I want to make sure it can boot up giving the complexity in dlls

wli3 commented on src/Resolvers/Microsoft.NET.Sdk.WorkloadManifestReader/SdkDirectoryWorkloadManifestProvider.cs:
```diff
@@ -51,7 +51,10 @@ public IEnumerable<Stream> GetManifests()
                 {
                     var workloadManifest = Path.Combine(workloadName, "WorkloadManifest.json");
-                    yield return File.OpenRead(workloadManifest);
+                    if (File.Exists(workloadManifest))
```
This will break one of my test. It make sense to throw. There has to be a WorkloadManifest.json under manifest folder right?

dsplaisted replied:
I reverted this change.

wli3 commented on src/Resolvers/Microsoft.NET.Sdk.WorkloadResolver/MSBuildWorkloadSdkResolver.cs:
```diff
@@ -0,0 +1,155 @@
+
+            var sdkDirectory = GetSdkDirectory(context);
+            //  TODO: Is the directory name OK, or should we read the .version file (we don't want to use the Cli.Utils library to read it on .NET Framework)
+            var sdkVersion = Path.GetFileName(sdkDirectory);
```
read the folder name is enough. It is a contract used by at least template engine already

wli3 commented on src/Resolvers/Microsoft.NET.Sdk.WorkloadResolver/MSBuildWorkloadSdkResolver.cs:
```diff
@@ -0,0 +1,155 @@
+                }
+                return factory.IndicateSuccess(autoImportSdkPaths, sdkReference.Version);
+            }
+            else if (sdkReference.Name == "TestSdk")
```
This is temporary? Add a hook for test, and we should do the rest in the test setup

dsplaisted replied:
Oops, yes, I've now removed this.

dsplaisted committed:
Code review feedback

dsplaisted committed:
Put workload resolution behind feature flag

dsplaisted committed:
Use MSBuild SDK resolver to import workload manifest targets

This allows the calculation of the manifest paths to be centralized in C# code instead of
duplicated in MSBuild XML.

dsplaisted committed:
Add localized error message for missing workloads

dsplaisted committed:
Use workload resolver to get workload pack paths

wli3 commented on src/Resolvers/Microsoft.DotNet.MSBuildSdkResolver/Microsoft.DotNet.MSBuildSdkResolver.csproj:
```diff
@@ -1,13 +1,19 @@
 
   <PropertyGroup>
-    <TargetFrameworks>netstandard2.0;net472</TargetFrameworks>
+    <TargetFrameworks>$(SdkTargetFramework);net472</TargetFrameworks>
```
Could you also make the same change for template locator and all the "resolvers" project? They were all netstandard on core.

wli3 commented on src/Resolvers/Microsoft.NET.Sdk.WorkloadManifestReader/IWorkloadResolver.cs:
```diff
@@ -10,6 +10,11 @@ public interface IWorkloadResolver
+        /// </summary>
+        /// <param name="packId">A workload pack ID</param>
+        /// <returns>Information about the workload pack, or null if the specified pack ID isn't found in the manifests</returns>
+        WorkloadResolver.PackInfo? TryGetPackInfo(string packId);
```
nit: I usually expect a method starts with "try" uses out var and returns a bool . GetPackInfo should be enough given null annotation.

dsplaisted replied:
I just renamed an existing method.  I think that the naming pattern may have changed with nullable reference types.

wli3 commented on src/Resolvers/Microsoft.NET.Sdk.WorkloadResolver/MSBuildWorkloadSdkResolver.cs:
```diff
@@ -0,0 +1,168 @@
+
+            if (!_enabled)
+            {
+                string sentinelPath = Path.Combine(Path.GetDirectoryName(typeof(MSBuildWorkloadSdkResolver).Assembly.Location), "EnableWorkloadResolver.sentinel");
```
EnableWorkloadResolver.sentinel would be in VS hive for VS scenario. Is that ok?

wli3 commented on src/Resolvers/Microsoft.NET.Sdk.WorkloadResolver/MSBuildWorkloadSdkResolver.cs:
```diff
@@ -0,0 +1,168 @@
+
+        public MSBuildWorkloadSdkResolver()
+        {
+            //  Put workload resolution behind a feature flag.
```
We should file a bug to remove the feature flag in the future. A big problem of feature flag is people always forget to remove them

dsplaisted replied:
Filed #13849

dsplaisted committed:
Rename workload SDK resolver to avoid namespace conflict

wli3 commented on src/Tasks/Microsoft.NET.Build.Tasks/targets/Microsoft.NET.Sdk.props:
```diff
@@ -116,6 +116,17 @@ Copyright (c) .NET Foundation. All rights reserved.
+  <!-- Put workload resolution behind a feature flag.  It can be enabled either by setting the MSBuildEnableWorkloadResolver environment variable to true, or by
+       putting an EnableWorkloadResolver.sentinel file beside the MSBuild SDK resolver DLL -->
+  <PropertyGroup Condition="'$(MSBuildEnableWorkloadResolver)' == ''">
+    <__EnableWorkloadResolverSentinelPath Condition="'$(MSBuildRuntimeType)' == 'Core'">$(MSBuildBinPath)\EnableWorkloadResolver.sentinel</__EnableWorkloadResolverSentinelPath>
```
Tested the feature flag works for on/off * commandline/VS * environment/sentinel, 8 scenarios?

wli3 commented on src/Resolvers/Microsoft.NET.Sdk.WorkloadResolver/MSBuildWorkloadSdkResolver.cs:
```diff
@@ -0,0 +1,168 @@
+
+            InitializeWorkloadResolver(resolverContext);
+
+            if (sdkReference.Name.Equals("Microsoft.NET.SDK.WorkloadAutoImportPropsLocator", StringComparison.OrdinalIgnoreCase))
```
Does the test cover workload props and targets can be included to the project?

dsplaisted replied:
Yes, `It_should_build_with_workload` tests that the targets are imported by using a fake target platform that the workload targets enable.  And `It_should_import_AutoImports_for_installed_workloads` checks that the props are imported via a property that is set in the .props file.

wli3 reviewed (commented):
And there are few todos.

dsplaisted committed:
Fix testing with resolver on full framework

dsplaisted committed:
Use stage 2 MSBuild SDK resolver for full Framework tests

dsplaisted committed:
Fix compiling on non-Windows

dsplaisted committed:
Retarget workload / resolver projects to net472

dsplaisted committed:
Fix error packing SDK Resolver

dsplaisted committed:
Disable workload tests on full framework

dsplaisted committed:
Fix casing mismatch in test workload
